### PR TITLE
Create remoter::client wrapper

### PR DESCRIPTION
**Pull request by marcellp**
# Background
#### Link to issue 
https://biomage.atlassian.net/browse/BIOMAGE-503

#### Link to staging deployment URL 
N/A

#### Links to any Pull Requests related to this
N/A

#### Anything else the reviewers should know about the changes here
There have been two other cross-repo changes made:

* In [iac](https://github.com/biomage-ltd/iac/commit/13a71b7fef1a116e1221f1c4c69c1c456a1a6166), where a new path was created (/v0/pipelineResults), which will be used to send all data regarding the success/failure of pipeline steps.
* In [worker](https://github.com/biomage-ltd/worker/commit/62aad0d995322a387f4ea527839e33a4029db7f6), where a message attribute `type` was added to differentiate between WorkResponse and PipelineResponse type messages, both of which will use the same SNS topic.

Essentially, the SNS topic we currently use to route published messages to the API will be used to route pipeline success/failure messages as well. In order to make sure that the two different types of messages go to two different paths on the API, a message attribute had to be added to both the worker and this script, called `type`. For the worker, its value is `WorkResponse`, whereas for the pipeline it is `PipelineResponse`. As seen in the iac change, depending on this type parameter, responses are routed to different URLs in the API, which we will use to our advantage in another ticket.

# Changes
### Code changes
Added a wrapper function for running pipeline steps.

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [x] Tested locally with Inframock
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [x] Relevant Github READMEs updated
- [x] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

**Comment by ivababukova** [comment]
Regarding this part of the comment:

```
In iac, where a new path was created (/v0/pipelineResults), which will be used to send all data regarding the success/failure of pipeline steps.
In worker, where a message attribute type was added to differentiate between WorkResponse and PipelineResponse type messages, both of which will use the same SNS topic.
```

Why were there no merge requests opened for those?

File: remoter-client/src/wrapper.r [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    assign("scdata", data, pos = ".GlobalEnv")
+
+    # send result to API
+    message_id <- send_output_to_api(input, rest_of_results)
+
+    return(message_id)
+}
+
+wrapper(
+    "{'experimentId': \"5928a56c7cbff9de78974ab50765ed20\", 'taskName': \"test_fn\", 'config': {'auto': true, 'enabled': true, 'name': \"a\", 'limit': 200}}"
```
Comment (ivababukova) [inline remoter-client/src/wrapper.r:99]: I assume that this experiment id in here is hardcoded only temporary? How are you going to pass it to the wrapper script -- env variable?

File: remoter-client/src/wrapper.r [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    result <- sns$publish(
+        Message = RJSONIO::toJSON(
+            list(
+                input = input,
+                config = config,
+                plot_data = plot_data
+            )
+        ),
+        TopicArn = paste(
+            "arn:aws:sns:eu-west-1:242905224710",
```
Comment (ivababukova) [inline remoter-client/src/wrapper.r:49]: this sounds like a config variable

File: remoter-client/src/wrapper.r [hunk 3]
```diff
@@ -0,0 +1,100 @@
+require("RJSONIO")
+require("paws")
+require("zeallot")
+
+reload_from_s3 <- function(experiment_id) {
+    s3 <- paws::s3()
+    bucket_name <- paste("biomage-source", "staging", sep = "-")
```
Comment (ivababukova) [inline remoter-client/src/wrapper.r:7]: the bucket name should be set in the configuration file

File: remoter-client/src/test_fn.r [hunk 4]
```diff
@@ -0,0 +1,21 @@
+some_other_stuff <- function(a) {
+    a + 5
+}
+
+task <- function(input_data, input_config) {
+    config <- input_config
+
+    config$limit <- 202
```
Comment (ivababukova) [inline remoter-client/src/test_fn.r:8]: what is this limit for? Why testing with 202 is a good choice?

File: remoter-client/src/wrapper.r [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    return(obj)
+}
+
+
+run_step <- function(task_name, scdata, config) {
+    switch(task_name,
+        test_fn = {
+            import::from("test_fn.r", task)
+        },
+        stop(paste("Invalid task name given:", task_name))
```
Comment (ivababukova) [inline remoter-client/src/wrapper.r:24]: why do we call the test function in here?

File: remoter-client/src/wrapper.r [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    out <- task(scdata, config)
+    return(out)
+}
+
+send_output_to_api <- function(input, output) {
+    c(config, plot_data = plotData) %<-% output
+
+    cluster_env <- Sys.getenv("CLUSTER_ENV", "staging")
+    sandbox_id <- Sys.getenv("SANDBOX_ID", "demo")
+    topic_name <- paste("work-results", cluster_env, sandbox_id, sep = "-")
```
Comment (ivababukova) [inline remoter-client/src/wrapper.r:36]: similarly, `work-results` is the name of the topic where we place the results and has to be saved in the configuration file and used from there

**Comment by marcellp** [comment]
> Regarding this part of the comment:
> 
> ```
> In iac, where a new path was created (/v0/pipelineResults), which will be used to send all data regarding the success/failure of pipeline steps.
> In worker, where a message attribute type was added to differentiate between WorkResponse and PipelineResponse type messages, both of which will use the same SNS topic.
> ```
> 
> Why were there no merge requests opened for those?

There was no way to test it working otherwise. Most of these were small changes that did not change the code base, only the infrastructure as code bits. The actual handler for the API to receive updates and other such things will be relevant code changes which will be submitted as a merge request.

File: remoter-client/src/wrapper.r (hunk 5 above)
Comment (marcellp) [reply remoter-client/src/wrapper.r:24]: Correct. `test_fn` is just a dummy function that we can use to test the thing working properly. As bioinformaticians submit more code, they will have to add their tasks to this switch.

File: remoter-client/src/test_fn.r (hunk 4 above)
Comment (marcellp) [reply remoter-client/src/test_fn.r:8]: There is no point to it, it was just to see that R's copy-on-write works properly. I can just annotate this entire function as being just a dummy and it only serves as an example.

**Review by ivababukova — APPROVED** [review]
looks good, assuming that you have showed it and talked about it with Sebastian and Juanlu and you are all on the same page in terms of the implementation plan for the individual functions.

**Pull request merged**